src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,770 @@
+            env(jv);
+            env.close();
+
+            // Subject can delete non-expired
+            env(credentials::deleteCred(subject, subject, issuer, credType));
+            env.close();
+            {
+                auto const credKey =
+                    credentialKeylet(subject, issuer, credType);
+                BEAST_EXPECT(!env.le(credKey));
+                BEAST_EXPECT(!ownerCount(env, subject));
+                BEAST_EXPECT(!ownerCount(env, issuer));
+                auto const jle =
+                    credentials::ledgerEntry(env, subject, issuer, credType);
+                BEAST_EXPECT(
+                    jle.isObject() && jle.isMember(jss::result) &&
+                    jle[jss::result].isMember(jss::error) &&
+                    jle[jss::result][jss::error] == "entryNotFound");
+            }
+
+            jv = credentials::create(subject, issuer, credType);
+            jv[sfExpiration.jsonName] = t + 1000;
+            env(jv);
+            env.close();
+
+            // Issuer can delete non-expired
+            env(credentials::deleteCred(issuer, subject, issuer, credType));
+            env.close();
+            {
+                auto const credKey =
+                    credentialKeylet(subject, issuer, credType);
+                BEAST_EXPECT(!env.le(credKey));
+                BEAST_EXPECT(!ownerCount(env, subject));
+                BEAST_EXPECT(!ownerCount(env, issuer));
+                auto const jle =
+                    credentials::ledgerEntry(env, subject, issuer, credType);
+                BEAST_EXPECT(
+                    jle.isObject() && jle.isMember(jss::result) &&
+                    jle[jss::result].isMember(jss::error) &&
+                    jle[jss::result][jss::error] == "entryNotFound");
+            }
+        }
+    }
+
+    void
+    testCreateFailed(FeatureBitset features)
+    {
+        using namespace test::jtx;
+
+        const char credType[] = "abcde";
+
+        Account const issuer{"issuer"};
+        Account const subject{"subject"};
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            env.fund(XRP(5000), subject, issuer);
+            env.close();
+
+            {
+                testcase("Credentials fail, no subject param.");
+                auto jv = credentials::create(subject, issuer, credType);
+                jv.removeMember(jss::Subject);
+                env(jv, ter(temMALFORMED));
+            }
+
+            {
+                auto jv = credentials::create(subject, issuer, credType);
+                jv[jss::Subject] = to_string(xrpAccount());
+                env(jv, ter(temMALFORMED));
+            }
+
+            {
+                testcase("Credentials fail, no credentialType param.");
+                auto jv = credentials::create(subject, issuer, credType);
+                jv.removeMember(sfCredentialType.jsonName);
+                env(jv, ter(temMALFORMED));
+            }
```

Comment:
These errors won't even reach transaction processing due to the model verification, are these accurate tests?

Reply:
You typically get something like this (standalone):
```
{
    deprecated: "Signing support in the 'submit' command has been deprecated and will be removed in a future version of the server. Please migrate to a standalone signing tool.",
    error: 'invalidParams',
    error_code: 31,
    error_message: "Missing field 'tx_json.Subject'.",
    id: 1,
    request: {
      command: 'submit',
      id: 1,
      secret: '<masked>',
      tx_json: {
        ...
      }
    },
    status: 'error',
    type: 'response'
  }
```

Reply:
Ye, this is unit tests, and they can reach some internals of the transactions.  They are accurate as they are passed.